test_code.py

```diff
@@ -121,3 +125,73 @@ def test_55_swap_all_elements_if_reversed(self):
             "1,9 2,9 3,8 4,9 5,8 6,7 5,6 7,9 8,9", 
             "Many Elements get swapped around in a reversed list.")
 
+
+#testeing dynprog (refactored l8r ?)            
+            
+class TestDynProg(unittest.TestCase):
+    """Test dynanic programming task according to its specification.
+    """
+    
+    globals = None # will be set by test_code
+    
+    # set up & tear down
+    
+    def setUp(self):
+        """Initialize the test case."""        
+        self.old_stdout = sys.stdout
+        self.stdout = sys.stdout = StringIO()
+
+    def tearDown(self):
+        sys.stdout = self.old_stdout
+
+    # helpers
+
+    @property
+    def createChart(self):
+        createChart = self.globals.get("createChart", None)
+        self.assertIsNotNone(createChart, "Ich benötige eine Funktion namens createChart")
+        return createChart
+
+    @property
+    def bestChoice(self):
+        bestChoice = self.globals.get("bestChoice", None)
+        self.assertIsNotNone(bestChoice, "Ich benötige eine Funktion namens bestChoice")
+        return bestChoice
+         
+
+    # test case for optimization
+
+    def test_best_value(self):
+        items = [(3,4),(1,1),(4,5),(3,4),(2,2)]
+        maxWeight = 8
+        bestValRef = 7
+        self.assertEqual(self.createChart(items, maxWeight)[-1][-1][0],bestValRef, "Der bestmögliche Wert ist ein anderer")
+        
+    def test_best_choice(self):
+        items = [(3,4),(1,1),(4,5),(3,4),(2,2)]
+        maxWeight = 8
+        bestValRef = 7
+        chartRef = [[(0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0)],
+                    [(0, 0), (0, 0), (0, 0), (0, 0), (3, 1), (3, 1), (3, 1), (3, 1), (3, 1)],
+                    [(0, 0), (1, 1), (1, 1), (1, 1), (3, 0), (4, 1), (4, 1), (4, 1), (4, 1)],
+                    [(0, 0), (1, 0), (1, 0), (1, 0), (3, 0), (4, 0), (5, 1), (5, 1), (5, 1)],
+                    [(0, 0), (1, 0), (1, 0), (1, 0), (3, 0), (4, 0), (5, 0), (5, 0), (6, 1)],
+                    [(0, 0), (1, 0), (2, 1), (3, 1), (3, 0), (4, 0), (5, 0), (6, 1), (7, 1)]]
+        givenChoice = self.bestChoice(chartRef, items)
+        #items should be as long as bestChoice return value
+        self.assertEqual(len(givenChoice),len(items), "Die Länge des Rückgabewertes von bestChoice ist nicht der Spezification entsprechend")
+        #items should only contain 0 or 1
+        self.assertCountEqual(list(set(givenChoice)),[0,1],"bestChoice return value enthält andere Werte als 0 und 1")
+        #bestchoice meigt not be unique, so check value sum of given choice
+        valSum = 0
+        weightSum = 0
+        for i in range(len(items)):
```

Comment:
Diesen Test-Case kann man in viele Tests aufspalten.
Meine Regel:
- Wenn nach einem Assert nochmal eine berechnung kommt, dann kann der test-Case aufgespalten werden
- Am besten gitb es feedback, wenn der Test nur ein assert hat

Aus der Sicht, derer, die die test-cases lösen ist es dann so: Sie machen inkrementell mehr Tests grün, indem sie schrittweise das Ergebnis verbessern. Wenn sich nur die  Fehlermeldung ändert, kann es sein, dass sie nicht wissen, ob sie sich vor oder zurück bewegen.